### DISCUSSION TITLE
How do I customize the on_move_async previewer?

### DISCUSSION THREAD

**Discussion by comfortablynick** [liuchengxu/vim-clap#757] [Q&A] [answered]
I am trying to use the `on_move_async` api but I can't seem to make it work. For a simple example, this works without icons, but when `g:clap_enable_icon = v:true`, it does nothing. How do I manipulate the paths that are passed to `clap#client#call_preview_file()`?

**Comment by liuchengxu** [comment, marked as answer]
The backend requires a valid file path string. Given that clap is unable to interpret each entry and extract the filepath on its own, if your custom provider has a different format for each entry, you need to firstly extract a valid file path by hand and then pass it via `fpath` in the `clap#client#call_preview_file(extra)`. For instance:

```vim
let g:clap_provider_quick_open = {
    \ 'source':  ['0:~/.vimrc', '1:~/.bashrc', '2:~/.zshrc'],
    \ 'sink':    'e',
    \ 'on_move_async': {->clap#client#call_preview_file({'fpath': split(g:clap.display.getcurline(), ':')[1]})},
    \ }
```

This commit https://github.com/liuchengxu/vim-clap/commit/84cbe8dc7bfcc53422971ae89f1eda4d95611d50 is required for passing the entries that contain `~`, or you can convert it to an absolute path on Vim end.

**Reply by comfortablynick** [reply]
Thanks! I tried that before and it didn't work, but now it does.
